File: explore/string/add_binary.py
# Given two binary strings, return their sum.
class Solution:
    def addBinary(self, a: str, b: str) -> str:
        carry = 0
        result = ''

        a = list(a)
        b = list(b)

        # iterate through each binary string until they ALL fail.
        while a or b or carry:
            if a:
                carry += int(a.pop())
            if b:
                carry += int(b.pop())

            # if a and b binary place is 2, then we expect 0 in that place, 1 otherwise
            result += str(carry % 2)
            carry //= 2  # floor division

        return result[::-1]  # start : end : step -> reverses the string

File: golang/add_binary.go
func addBinary(a string, b string) string {
	i, j := len(a)-1, len(b)-1
	carry := 0
	result := ""

	for i >= 0 && j >= 0 {
		x, _ := strconv.Atoi(string(a[i]))
		y, _ := strconv.Atoi(string(b[j]))

		digit := strconv.Itoa((x + y + carry) % 2)
		carry = (x + y + carry) / 2
		result = digit + result

		i--
		j--
	}

	for i >= 0 {
		x, _ := strconv.Atoi(string(a[i]))

		digit := strconv.Itoa((x + carry) % 2)
		carry = (x + carry) / 2
		result = digit + result

		i--
	}
	for j >= 0 {
		y, _ := strconv.Atoi(string(b[j]))

		digit := strconv.Itoa((y + carry) % 2)
		carry = (y + carry) / 2
		result = digit + result

		j--
	}

	if carry == 1 {
		result = "1" + result
	}

	return result
}